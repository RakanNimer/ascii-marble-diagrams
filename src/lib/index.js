const marbleTypes = {
  '[': {
    type: 'OPERATOR_START'
  },
  ']': {
    type: 'OPERATOR_END'
  },
  '-': {
    type: 'EMPTY_FRAME'
  },
  '|': {
    type: 'COMPLETE'
  },
  '#': {
    type: 'ERROR'
  },
  '(': {
    type: 'GROUP_START'
  },
  ')': {
    type: 'GROUP_END'
  },
  '^': {
    type: 'SUBSCRIPTION_POINT'
  }
}
const parseMarbleDiagram = marbleString => {
  const marbleDiagram = Array.from(marbleString)
  const allCharacters = marbleDiagram.map((char, i) => {
    if (char in marbleTypes) {
      return Object.assign({}, marbleTypes[char], { index: i, char })
    }
    if (char.match(/[a-zA-Z0-9]/)) {
      return { type: 'VARIABLE', char, index: i }
    }
    console.log({ var: char })
    return { type: 'UNKNOWN', char, index: i }
  })
  return allCharacters
}

const tokenizeMarbleDiagram = marbleCharacters => {
  let groupStartIndex = -1
  let groups = {}
  let operatorStartIndex = -1
  let operators = {}

  return marbleCharacters.reduce((acc, current, i) => {
    if (groupStartIndex !== -1 && current.type !== 'GROUP_END') {
      groups[groupStartIndex] = [...groups[groupStartIndex], current]
      return acc
    }
    if (operatorStartIndex !== -1 && current.type !== 'OPERATOR_END') {
      operators[operatorStartIndex] = [
        ...operators[operatorStartIndex],
        current
      ]
      return acc
    }
    switch (current.type) {
      case 'OPERATOR_START': {
        operatorStartIndex = i
        operators[operatorStartIndex] = []
        return acc
      }
      case 'OPERATOR_END': {
        if (operatorStartIndex === -1) {
          console.error(
            `Operator not defined correctly. Failing silently. Error at index ${i}`
          )
          return acc
        }
        const operatorName = operators[operatorStartIndex]
        const token = {
          type: 'OPERATOR',
          char: operatorName,
          startIndex: operatorStartIndex,
          endIndex: i,
          index: operatorStartIndex
        }
        operatorStartIndex = -1
        return [...acc, token]
      }
      case 'GROUP_START': {
        groupStartIndex = i
        groups[groupStartIndex] = []
        return acc
      }
      case 'GROUP_END': {
        if (groupStartIndex === -1) {
          console.error(
            `Sync Grouped emission not tokenized correctly. Failing silently. Error at index ${i}`
          )
          return acc
        }
        const groupContent = groups[groupStartIndex]
        const token = {
          type: 'GROUP',
          char: groupContent,
          startIndex: groupStartIndex,
          endIndex: i,
          index: groupStartIndex
        }
        groupStartIndex = -1
        return [...acc, token]
      }
      default: {
        return [...acc, current]
      }
    }
  }, [])
}

export { parseMarbleDiagram, tokenizeMarbleDiagram }
