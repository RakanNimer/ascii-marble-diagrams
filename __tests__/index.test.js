import cases from 'jest-in-case'
import { parseMarbleDiagram, tokenizeMarbleDiagram } from '../src/lib/'

const diagrams = [
  { name: 'empty', diagram: '-' },
  { name: 'one', diagram: '-a-' },
  { name: 'two consecutive', diagram: '-cb-' },
  { name: 'with error', diagram: '-b-#-' },
  { name: 'hot', diagram: '-^--b--' },
  { name: 'with sync group', diagram: '-(bcd)-' },
  { name: 'with unexpected character', diagram: '->-' },
  { name: 'operator', diagram: '[switchMap]' },
  { name: 'invalid operator', diagram: 'switchMap]' },
  { name: 'operator & expression', diagram: '[filter]{i => i < 3}' },
  { name: 'invalid expression', diagram: 'i => i < 3}' }
]

describe('parseMarbleDiagram', () => {
  test('exports', () => {
    expect(parseMarbleDiagram).toBeDefined()
  })
  cases(
    'parses marble diagrams',
    opts => {
      const { diagram } = opts
      const parsedDiagram = parseMarbleDiagram(diagram)
      expect(parsedDiagram).toMatchSnapshot()
    },
    diagrams
  )
})

describe('tokenizeMarbleDiagram', () => {
  test('exports', () => {
    expect(tokenizeMarbleDiagram).toBeDefined()
  })
  cases(
    'tokenizes parsed marble diagrams',
    opts => {
      const { diagram } = opts
      const parsedDiagram = parseMarbleDiagram(diagram)
      const diagramTokens = tokenizeMarbleDiagram(parsedDiagram)
      expect(diagramTokens).toMatchSnapshot()
      // console.log({ parsedDiagram, diagramTokens })
    },
    diagrams
  )
})
