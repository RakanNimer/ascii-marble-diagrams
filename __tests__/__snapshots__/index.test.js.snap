// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseMarbleDiagram parses marble diagrams empty 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams hot 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "^",
    "index": 1,
    "type": "SUBSCRIPTION_POINT",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "-",
    "index": 3,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "b",
    "index": 4,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 5,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "-",
    "index": 6,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams invalid operator 1`] = `
Array [
  Object {
    "char": "s",
    "index": 0,
    "type": "VARIABLE",
  },
  Object {
    "char": "w",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "i",
    "index": 2,
    "type": "VARIABLE",
  },
  Object {
    "char": "t",
    "index": 3,
    "type": "VARIABLE",
  },
  Object {
    "char": "c",
    "index": 4,
    "type": "VARIABLE",
  },
  Object {
    "char": "h",
    "index": 5,
    "type": "VARIABLE",
  },
  Object {
    "char": "M",
    "index": 6,
    "type": "VARIABLE",
  },
  Object {
    "char": "a",
    "index": 7,
    "type": "VARIABLE",
  },
  Object {
    "char": "p",
    "index": 8,
    "type": "VARIABLE",
  },
  Object {
    "char": "]",
    "index": 9,
    "type": "OPERATOR_END",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams one 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "a",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams operator 1`] = `
Array [
  Object {
    "char": "[",
    "index": 0,
    "type": "OPERATOR_START",
  },
  Object {
    "char": "s",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "w",
    "index": 2,
    "type": "VARIABLE",
  },
  Object {
    "char": "i",
    "index": 3,
    "type": "VARIABLE",
  },
  Object {
    "char": "t",
    "index": 4,
    "type": "VARIABLE",
  },
  Object {
    "char": "c",
    "index": 5,
    "type": "VARIABLE",
  },
  Object {
    "char": "h",
    "index": 6,
    "type": "VARIABLE",
  },
  Object {
    "char": "M",
    "index": 7,
    "type": "VARIABLE",
  },
  Object {
    "char": "a",
    "index": 8,
    "type": "VARIABLE",
  },
  Object {
    "char": "p",
    "index": 9,
    "type": "VARIABLE",
  },
  Object {
    "char": "]",
    "index": 10,
    "type": "OPERATOR_END",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams two consecutive 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "c",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "b",
    "index": 2,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 3,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams with error 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "b",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "#",
    "index": 3,
    "type": "ERROR",
  },
  Object {
    "char": "-",
    "index": 4,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams with sync group 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "(",
    "index": 1,
    "type": "GROUP_START",
  },
  Object {
    "char": "b",
    "index": 2,
    "type": "VARIABLE",
  },
  Object {
    "char": "c",
    "index": 3,
    "type": "VARIABLE",
  },
  Object {
    "char": "d",
    "index": 4,
    "type": "VARIABLE",
  },
  Object {
    "char": ")",
    "index": 5,
    "type": "GROUP_END",
  },
  Object {
    "char": "-",
    "index": 6,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`parseMarbleDiagram parses marble diagrams with unexpected character 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "{",
    "index": 1,
    "type": "UNKNOWN",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams empty 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams hot 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "^",
    "index": 1,
    "type": "SUBSCRIPTION_POINT",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "-",
    "index": 3,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "b",
    "index": 4,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 5,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "-",
    "index": 6,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams invalid operator 1`] = `
Array [
  Object {
    "char": "s",
    "index": 0,
    "type": "VARIABLE",
  },
  Object {
    "char": "w",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "i",
    "index": 2,
    "type": "VARIABLE",
  },
  Object {
    "char": "t",
    "index": 3,
    "type": "VARIABLE",
  },
  Object {
    "char": "c",
    "index": 4,
    "type": "VARIABLE",
  },
  Object {
    "char": "h",
    "index": 5,
    "type": "VARIABLE",
  },
  Object {
    "char": "M",
    "index": 6,
    "type": "VARIABLE",
  },
  Object {
    "char": "a",
    "index": 7,
    "type": "VARIABLE",
  },
  Object {
    "char": "p",
    "index": 8,
    "type": "VARIABLE",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams one 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "a",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams operator 1`] = `
Array [
  Object {
    "char": Array [
      Object {
        "char": "s",
        "index": 1,
        "type": "VARIABLE",
      },
      Object {
        "char": "w",
        "index": 2,
        "type": "VARIABLE",
      },
      Object {
        "char": "i",
        "index": 3,
        "type": "VARIABLE",
      },
      Object {
        "char": "t",
        "index": 4,
        "type": "VARIABLE",
      },
      Object {
        "char": "c",
        "index": 5,
        "type": "VARIABLE",
      },
      Object {
        "char": "h",
        "index": 6,
        "type": "VARIABLE",
      },
      Object {
        "char": "M",
        "index": 7,
        "type": "VARIABLE",
      },
      Object {
        "char": "a",
        "index": 8,
        "type": "VARIABLE",
      },
      Object {
        "char": "p",
        "index": 9,
        "type": "VARIABLE",
      },
    ],
    "endIndex": 10,
    "index": 0,
    "startIndex": 0,
    "type": "OPERATOR",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams two consecutive 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "c",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "b",
    "index": 2,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 3,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams with error 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "b",
    "index": 1,
    "type": "VARIABLE",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "#",
    "index": 3,
    "type": "ERROR",
  },
  Object {
    "char": "-",
    "index": 4,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams with sync group 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": Array [
      Object {
        "char": "b",
        "index": 2,
        "type": "VARIABLE",
      },
      Object {
        "char": "c",
        "index": 3,
        "type": "VARIABLE",
      },
      Object {
        "char": "d",
        "index": 4,
        "type": "VARIABLE",
      },
    ],
    "endIndex": 5,
    "index": 1,
    "startIndex": 1,
    "type": "GROUP",
  },
  Object {
    "char": "-",
    "index": 6,
    "type": "EMPTY_FRAME",
  },
]
`;

exports[`tokenizeMarbleDiagram tokenizes parsed marble diagrams with unexpected character 1`] = `
Array [
  Object {
    "char": "-",
    "index": 0,
    "type": "EMPTY_FRAME",
  },
  Object {
    "char": "{",
    "index": 1,
    "type": "UNKNOWN",
  },
  Object {
    "char": "-",
    "index": 2,
    "type": "EMPTY_FRAME",
  },
]
`;
